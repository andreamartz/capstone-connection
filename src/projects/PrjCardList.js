import React, { useState, useEffect, useContext } from "react";
import Masonry from 'react-masonry-css';
import Box from '@material-ui/core/Box';
import Container from '@material-ui/core/Container';
import { makeStyles, useTheme } from '@material-ui/core/styles';
import useMediaQuery from '@material-ui/core/useMediaQuery';
import CapConApi from "../api/api";
import LoadingSpinner from "../common/LoadingSpinner";
import PrjCardVert from './PrjCardVert';
import SearchForm from './SearchForm';
import SortForm from './SortForm';
import UserContext from "../auth/UserContext";
import { asyncWrapper, toggleLikeProject } from "../utils";
import "./PrjCardList.css";

const useStyles = makeStyles((theme) => ({
  project: {
    marginTop: theme.spacing(3),
    marginBottom: theme.spacing(3)
  },
  searchForm: {
    flexGrow: '1'
  }
}));

/** Show project cards.
 * 
 * On mount, loads projects from API:
 *   * loads all projects if no userId
 *   * loads user's projects if userId
 *  
 */ 

const PrjCardList = ({ userId }) => {
  const { currentUser } = useContext(UserContext);
  const [projects, setProjects] = useState([]);
  const [searchTerm, setSearchTerm] = useState('');
  const [sortVariable, setSortVariable] = useState('');
  const [showSearchAndSort, setShowSearchAndSort] = useState(false);

  const classes = useStyles();

  useEffect(() => {
    async function getAllProjectsOnMount() {
      const projects = userId 
        ? (await CapConApi.getUserProjects(userId))
        : (await CapConApi.getProjects(searchTerm, sortVariable));
      setProjects(projects);
      if (userId) {
        setShowSearchAndSort(false);
      } else {
        setShowSearchAndSort(true);
      }
    }
    getAllProjectsOnMount();
  }, [userId, sortVariable, searchTerm]); 

  // Breakpoints
  const theme = useTheme();
  const isSmallScreen = useMediaQuery(theme.breakpoints.down('sm'));
  const flexDirection = isSmallScreen ? "column" : "row";
  const justifyContent = isSmallScreen ? "center" : "space-between";
  const alignItems = isSmallScreen ? "flex-start" : "space-between";
  const marginBottom = isSmallScreen ? 10 : 1;

  const search = async () => {
    const projects = await CapConApi.getProjects(searchTerm, sortVariable);
    setProjects(projects);
  };

  const sort = async (sortVariable) => {
    console.log("SORTVARIABLE: ", sortVariable, "PROJECTS BEFORE SORT: ", projects);
    const sortedProjects = await CapConApi.getProjects(null, sortVariable);
    console.log("PROJECTS AFTER SORT: ", sortedProjects);
    setProjects(sortedProjects);
  }

  if (!projects) return <LoadingSpinner />;

  const breakpoints = {
    default: 3,  // default number of columns
    960: 2,     // at screen width of 1100px, reduce to 2 columns
    600: 1
  }

  console.log("PROJECTS STATE: ", projects);

  return (
    <Container>
      <Box display='flex'
        flexDirection={flexDirection}
        justifyContent={justifyContent}
        alignItems={alignItems}
        mb={5}
      >
        {showSearchAndSort && <SearchForm className={classes.searchForm} search={search} searchTerm={searchTerm} setSearchTerm={setSearchTerm} />}
        <Box mb={marginBottom}>
          {showSearchAndSort && <SortForm sort={sort} sortVariable={sortVariable} setSortVariable={setSortVariable}/>}
        </Box>
      </Box>

      {projects.length 
        ? (
          <Masonry
            breakpointCols={breakpoints}
            className="masonry-grid"
            columnClassName="masonry-grid_column"
          >
            {projects.map((p, idx) => (
              <div key={p.id} className={classes.project}>
                <PrjCardVert
                  project={p}
                  toggleLikeProject={() => {toggleLikeProject(idx, currentUser, projects, setProjects)}}
                />
              </div>
            ))}
          </Masonry>
        ) : (
          <p>Sorry, no projects were found!</p>
        )
      }    
    </Container>
  );
}

export default PrjCardList;