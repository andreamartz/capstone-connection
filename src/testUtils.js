import React from "react";
import UserContext from "./auth/UserContext";

const demoUser = {
  id: 1,
  username: "testuser",
  firstName: "testfirst",
  lastName: "testlast",
  bio: "This is my bio",
  email: "test@test.net",
  photoUrl: null,
};

const UserProvider =
    ({ children, currentUser = demoUser }) => (
    <UserContext.Provider value={{ currentUser }}>
      {children}
    </UserContext.Provider>
);

export { UserProvider };
