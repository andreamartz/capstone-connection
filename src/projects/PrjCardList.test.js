import React from 'react';
import { render } from '@testing-library/react';
import PrjCardList from './PrjCardList';
import { UserProvider } from '../testUtils';

test('it renders without crashing', function () {
	render(
		<UserProvider>
			<PrjCardList />
		</UserProvider>,
	);
});

test('it matches snapshot', function () {
	const { asFragment } = render(
		<UserProvider>
			<PrjCardList />
		</UserProvider>,
	);
	expect(asFragment()).toMatchSnapshot();
});
