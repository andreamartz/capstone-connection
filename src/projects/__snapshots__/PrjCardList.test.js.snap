// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiContainer-root MuiContainer-maxWidthLg"
  >
    <div
      class="MuiBox-root MuiBox-root-7"
    >
      <div
        class="MuiBox-root MuiBox-root-8"
      />
    </div>
    <p>
      Sorry, no projects were found!
    </p>
  </div>
</DocumentFragment>
`;
